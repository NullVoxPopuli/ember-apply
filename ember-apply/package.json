{
  "name": "ember-apply",
  "version": "2.2.1",
  "description": "automatic integration and configuration from the community",
  "repository": {
    "url": "https://github.com/NullVoxPopuli/ember-apply",
    "type": "https",
    "directory": "packages/ember-apply"
  },
  "main": "src/index.js",
  "exports": {
    ".": "./src/index.js",
    "./test-utils": "./src/test-utils.js"
  },
  "type": "module",
  "author": "NullVoxPopuli",
  "license": "MIT",
  "private": false,
  "bin": {
    "ember-apply": "src/cli/index.js"
  },
  "types": "./declarations/src/",
  "files": [
    "src/**/*",
    "declarations/**/*",
    "README.md"
  ],
  "scripts": {
    "prepack": "yarn build:types && cp ../README.md ./",
    "build:types": "tsc --build",
    "lint": "yarn lint:js",
    "lint:js": "eslint .",
    "lint:fix": "yarn lint:js --fix",
    "test": "vitest --coverage --no-watch --no-threads",
    "test:watch": "vitest --watch --no-threads"
  },
  "dependencies": {
    "@babel/core": "^7.16.7",
    "@babel/preset-env": "^7.16.8",
    "chalk": "^5.0.0",
    "ember-template-recast": "^6.1.2",
    "execa": "^6.0.0",
    "fs-extra": "^10.0.0",
    "jscodeshift": "^0.13.1",
    "latest-version": "^6.0.0",
    "ora": "^6.0.1",
    "pacote": "^13.0.0",
    "posthtml": "^0.16.5",
    "semver": "^7.3.5",
    "unified": "^10.1.1",
    "yargs": "^17.3.1",
    "yarn-workspaces-list": "^0.2.0"
  },
  "devDependencies": {
    "@nullvoxpopuli/eslint-configs": "^2.1.17",
    "@types/fs-extra": "^9.0.13",
    "@types/jscodeshift": "^0.11.3",
    "@types/node": "^17.0.8",
    "@types/pacote": "^11.1.3",
    "@types/semver": "^7.3.9",
    "@types/yargs": "^17.0.8",
    "@vitest/ui": "^0.6.0",
    "babel-eslint": "^10.1.0",
    "c8": "^7.11.0",
    "eslint": "^7.0.0",
    "eslint-plugin-decorator-position": "^4.0.1",
    "eslint-plugin-import": "^2.25.4",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-prettier": "^4.0.0",
    "prettier": "^2.5.1",
    "semantic-release": "^19.0.0",
    "semantic-release-monorepo": "^7.0.0",
    "typescript": "4.6.2",
    "vite": "^2.7.13",
    "vitest": "^0.5.0"
  },
  "engines": {
    "node": ">=16.0.0"
  },
  "volta": {
    "node": "16.14.0",
    "yarn": "1.22.17"
  }
}
