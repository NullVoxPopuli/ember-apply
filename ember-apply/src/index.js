// @ts-check
//
//
// Each of these re-exported modules points at an index.js file which
// defines the Public API for each namespace

export * as automation from './-private/automation/index.js';
export * as css from './-private/css/index.js';
export * as ember from './-private/ember/index.js';
export * as files from './-private/files/index.js';
export * as html from './-private/html/index.js';
export * as js from './-private/js/index.js';
export * as npm from './-private/npm/index.js';
export * as packageJson from './-private/package-json/index.js';
export * as project from './-private/project/index.js';
