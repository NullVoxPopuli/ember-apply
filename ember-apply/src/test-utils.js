import { execa } from 'execa';
import fs from 'fs/promises';
import os from 'os';
import path, { dirname } from 'path';
import { fileURLToPath } from 'url';

import { packageJson, project } from './index.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

export async function newTmpDir() {
  const tmpDir = await fs.mkdtemp(
    path.join(os.tmpdir(), `ember-apply-${Date.now()}-`),
  );

  return tmpDir;
}

export async function newEmberApp() {
  let dir = await newTmpDir();

  await execa('ember', ['-v'], { cwd: dir });
  await execa('ember', ['new', 'test-app', '--skip-npm'], {
    cwd: dir,
  });

  return path.join(dir, 'test-app');
}

export async function newEmberAddon() {
  let dir = await newTmpDir();

  await execa('ember', ['-v'], { cwd: dir });
  await execa('ember', ['addon', 'test-addon', '--skip-npm'], {
    cwd: dir,
  });

  return path.join(dir, 'test-addon');
}

/**
 * @param {string} appPath the app to apply the applyable to
 * @param {string} applyablePath the path to the applyable
 */
export async function apply(appPath, applyablePath) {
  let cli = path.join(__dirname, 'cli/index.js');
  let target = applyablePath.startsWith('@')
    ? applyablePath
    : path.resolve(applyablePath);

  await execa('node', [cli, '--verbose', target], {
    cwd: appPath,
    stdio: process.env.DEBUG ? 'inherit' : undefined,
  });
}

/**
 * @param {string} appPath the path to git diff
 */
export async function diffSummary(appPath) {
  let { stdout } = await execa('git', ['diff', '--compact-summary'], {
    cwd: appPath,
  });

  return stdout;
}

/**
 * @typedef {object} DiffOptions
 * @property {boolean} [ignoreVersions]
 *
 * @param {string} appPath the path to git diff
 * @param {DiffOptions} [options]
 */
export async function diff(appPath, options = {}) {
  let { stdout } = await execa('git', ['diff', '--name-only'], {
    cwd: appPath,
  });

  let files = stdout.split('\n');

  let diff = '';

  for (let filePath of files) {
    if (!filePath) {
      continue;
    }

    if (filePath.endsWith('package.json') && options?.ignoreVersions) {
      let fileBuffer = await fs.readFile(path.join(appPath, filePath));
      let fileString = fileBuffer.toString();
      let json = JSON.parse(fileString);

      for (let section of [
        'dependencies',
        'devDependencies',
        'peerDependencies',
        'engines',
        'volta',
      ]) {
        if (json[section]) {
          json[section] = Object.keys(json[section]);
        }
      }

      diff += `\n${filePath}\n${JSON.stringify(json, null, 2)}\n`;
      continue;
    }

    // Why is this not easier....
    // let normalCommand =
    //   `git diff ${filePath} ` +
    //   `| sed -n '/^---/!p' ` +
    //   `| sed -n '/^+++/!p' ` +
    //   `| sed -n '/^@@/!p' ` +
    //   `| sed -n '/^index /!p'`;

    let { stdout: fullDiff } = await execa(`git`, ['diff', filePath], {
      cwd: appPath,
    });

    let shortDiff = fullDiff
      .replace(/^---.+$/, '')
      .replace(/^\+\+\+.+$/, '')
      .replace(/index.+$/, '')
      .replace(/@@.+$/, '');

    diff += shortDiff + '\n';
  }

  return diff;
}

/**
 * Gets text content of the file all in one go
 * fs.readFile with the extra steps
 *
 * @param {string} filePath
 * @returns {Promise<string>} the file
 */
export async function readFile(filePath) {
  let file = await fs.readFile(filePath);

  return file.toString();
}

/**
 * Gets JSON content of the file all in one go
 * fs.readFile with the extra steps
 *
 * @param {string} filePath
 * @returns {Promise<object>} the file
 */
export async function readJson(filePath) {
  let text = await readFile(filePath);

  return JSON.parse(text);
}

/**
 * Creates a new monorepo with git in a tmp directory
 *
 * @param {string[]} workspaces list of workspace names
 */
export async function newMonorepo(workspaces) {
  let root = await newTmpDir();

  await execa('git', ['init'], { cwd: root });

  await fs.writeFile(
    path.join(root, 'package.json'),
    `{ "workspaces": [${workspaces.map((w) => `"${w}"`)}], "private": true }`,
  );
  await fs.writeFile(
    path.join(root, 'pnpm-workspace.yaml'),
    `packages:\n` + workspaces.map((w) => `- ${w}\n`).join(''),
  );

  for (let workspace of workspaces) {
    let safeName = workspace.replaceAll('/', '-');

    await fs.mkdir(path.join(root, `${workspace}`), { recursive: true });
    await fs.writeFile(
      path.join(root, `${workspace}/package.json`),
      `{ "name": "test-${safeName}" , "dependencies": {}, "version": "0.0.0" }`,
    );
  }

  return root;
}

/**
 * @param {string} root
 */
export async function readAllPackageJson(root) {
  let workspaces = await project.getWorkspaces(root);

  let manifests = [];

  // done serially, so order is consistent
  for (let workspace of workspaces) {
    let manifest = await packageJson.read(workspace);

    manifests.push(manifest);
  }

  return manifests;
}
