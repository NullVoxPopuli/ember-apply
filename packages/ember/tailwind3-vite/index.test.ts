import { apply, diff, diffSummary, newEmberApp } from "ember-apply/test-utils";
import { describe, expect, it } from "vitest";

import { default as tailwindVite } from "./index.js";

describe("tailwind3-vite", () => {
  it("default export exists", () => {
    expect(typeof tailwindVite).toEqual("function");
  });

  describe("applying to an ember app", () => {
    it("works via CLI", async () => {
      let appLocation = await newEmberApp([
        "--blueprint",
        "@embroider/app-blueprint",
      ]);

      await apply(appLocation, tailwindVite.path);

      expect(await diffSummary(appLocation)).toMatchSnapshot();
      expect(
        await diff(appLocation, { ignoreVersions: true })
      ).toMatchSnapshot();
    });
  });
});
