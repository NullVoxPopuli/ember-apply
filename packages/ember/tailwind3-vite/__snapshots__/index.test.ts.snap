// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tailwind3-vite > applying to an ember app > works via CLI 1`] = `
" app/app.js                | 2 ++
 app/styles/app.css        | 3 +++
 package.json              | 3 +++
 postcss.config.cjs (new)  | 8 ++++++++
 tailwind.config.cjs (new) | 9 +++++++++
 5 files changed, 25 insertions(+)"
`;

exports[`tailwind3-vite > applying to an ember app > works via CLI 2`] = `
"diff --git a/app/app.js b/app/app.js
index a3c0402..57368f0 100644
--- a/app/app.js
+++ b/app/app.js
@@ -4,6 +4,8 @@ import Resolver from 'ember-resolver';
 import loadInitializers from 'ember-load-initializers';
 import config from './config/environment';
 
+import './styles/app.css'
+
 export default class App extends Application {
   modulePrefix = config.modulePrefix;
   podModulePrefix = config.podModulePrefix;
diff --git a/app/styles/app.css b/app/styles/app.css
index 2763afa..192268f 100644
--- a/app/styles/app.css
+++ b/app/styles/app.css
@@ -1 +1,4 @@
 /* Ember supports plain CSS out of the box. More info: https://cli.emberjs.com/release/advanced-use/stylesheets/ */
+@tailwind base;
+@tailwind components;
+@tailwind utilities

package.json
{
  \\"name\\": \\"test-app\\",
  \\"version\\": \\"0.0.0\\",
  \\"private\\": true,
  \\"description\\": \\"Small description for test-app goes here\\",
  \\"repository\\": \\"\\",
  \\"license\\": \\"MIT\\",
  \\"author\\": \\"\\",
  \\"directories\\": {
    \\"doc\\": \\"doc\\",
    \\"test\\": \\"tests\\"
  },
  \\"scripts\\": {
    \\"build\\": \\"vite build\\",
    \\"lint\\": \\"concurrently \\\\\\"npm:lint:*(!fix)\\\\\\" --names \\\\\\"lint:\\\\\\" --prefixColors auto\\",
    \\"lint:css\\": \\"stylelint \\\\\\"**/*.css\\\\\\"\\",
    \\"lint:css:fix\\": \\"concurrently \\\\\\"npm:lint:css -- --fix\\\\\\"\\",
    \\"lint:fix\\": \\"concurrently \\\\\\"npm:lint:*:fix\\\\\\" --names \\\\\\"fix:\\\\\\" --prefixColors auto\\",
    \\"lint:hbs\\": \\"ember-template-lint .\\",
    \\"lint:hbs:fix\\": \\"ember-template-lint . --fix\\",
    \\"lint:js\\": \\"eslint . --cache\\",
    \\"lint:js:fix\\": \\"eslint . --fix\\",
    \\"start\\": \\"vite\\",
    \\"test\\": \\"concurrently \\\\\\"npm:lint\\\\\\" \\\\\\"npm:test:*\\\\\\" --names \\\\\\"lint,test:\\\\\\" --prefixColors auto\\",
    \\"test:ember\\": \\"vite build --mode test && ember test --path dist\\"
  },
  \\"devDependencies\\": [
    \\"@babel/core\\",
    \\"@babel/eslint-parser\\",
    \\"@babel/plugin-transform-runtime\\",
    \\"@babel/runtime\\",
    \\"@ember/optional-features\\",
    \\"@ember/string\\",
    \\"@ember/test-helpers\\",
    \\"@embroider/compat\\",
    \\"@embroider/config-meta-loader\\",
    \\"@embroider/core\\",
    \\"@embroider/test-setup\\",
    \\"@embroider/vite\\",
    \\"@eslint/js\\",
    \\"@glimmer/component\\",
    \\"@glimmer/tracking\\",
    \\"@rollup/plugin-babel\\",
    \\"autoprefixer\\",
    \\"babel-plugin-ember-template-compilation\\",
    \\"concurrently\\",
    \\"decorator-transforms\\",
    \\"ember-auto-import\\",
    \\"ember-cli\\",
    \\"ember-cli-babel\\",
    \\"ember-cli-htmlbars\\",
    \\"ember-data\\",
    \\"ember-load-initializers\\",
    \\"ember-modifier\\",
    \\"ember-page-title\\",
    \\"ember-qunit\\",
    \\"ember-resolver\\",
    \\"ember-source\\",
    \\"ember-template-imports\\",
    \\"ember-template-lint\\",
    \\"ember-welcome-page\\",
    \\"eslint\\",
    \\"eslint-config-prettier\\",
    \\"eslint-plugin-ember\\",
    \\"eslint-plugin-n\\",
    \\"eslint-plugin-prettier\\",
    \\"eslint-plugin-qunit\\",
    \\"globals\\",
    \\"loader.js\\",
    \\"postcss\\",
    \\"prettier\\",
    \\"prettier-plugin-ember-template-tag\\",
    \\"qunit\\",
    \\"qunit-dom\\",
    \\"stylelint\\",
    \\"stylelint-config-standard\\",
    \\"stylelint-prettier\\",
    \\"tailwindcss\\",
    \\"tracked-built-ins\\",
    \\"vite\\"
  ],
  \\"engines\\": [
    \\"node\\"
  ],
  \\"ember\\": {
    \\"edition\\": \\"octane\\"
  },
  \\"ember-addon\\": {
    \\"type\\": \\"app\\",
    \\"version\\": 2
  },
  \\"exports\\": {
    \\"./tests/*\\": \\"./tests/*\\",
    \\"./*\\": \\"./app/*\\"
  }
}
diff --git a/postcss.config.cjs b/postcss.config.cjs
new file mode 100644
index 0000000..5f067a0
--- /dev/null
+++ b/postcss.config.cjs
@@ -0,0 +1,8 @@
+module.exports = {
+  plugins: {
+    tailwindcss: {
+      config: \\"tailwind.config.cjs\\"
+    },
+    autoprefixer: {},
+  },
+}
diff --git a/tailwind.config.cjs b/tailwind.config.cjs
new file mode 100644
index 0000000..e5e907d
--- /dev/null
+++ b/tailwind.config.cjs
@@ -0,0 +1,9 @@
+/** @type {import('tailwindcss').Config} */
+module.exports = {
+  content: [\\"index.html\\", \\"app/**/*.{js,ts,hbs,gjs,gts,html}\\"],
+  theme: {
+    extend: {},
+  },
+  plugins: [],
+}
+
"
`;
